docs/architecture/grpc-for-wcf-developers/apm.md

```diff
@@ -1,11 +1,221 @@
+
+ASP.NET Core provides support for logging out-of-the-box, in the form of the [Microsoft.Extensions.Logging package](https://www.nuget.org/packages/Microsoft.Extensions.Logging). This is included with the Web SDK so there is no need to install it manually. The gRPC framework writes its internal diagnostic logging messages to ASP.NET Core logging so they can be processed/stored along with the application's own messages.
+
+### Producing log messages
+
+The Logging extension is automatically registered with ASP.NET Core's dependency injection system, so loggers can be specified as a constructor parameter on gRPC service types.
+
+```csharp
+public class StockData : Stocks.StocksBase
+{
+    private readonly ILogger<StockData> _logger;
+
+    public StockData(ILogger<StockData> logger)
+    {
+        _logger = logger;
+    }
+}
+```
+
+A lot of log messages around requests, exceptions, etc. are provided by the ASP.NET Core and gRPC framework components. You should add your own log messages to provide detail and context about application logic rather than lower level concerns.
+
+For more information please refer to the [ASP.NET Core Logging documentation](https://docs.microsoft.com/aspnet/core/fundamentals/logging/?view=aspnetcore-3.0).
+
+## Metrics in ASP.NET Core gRPC
+
+The .NET Core framework provides a set of components for measuring performance in the [System.Diagnostics.Tracing](https://www.nuget.org/packages/System.Diagnostics.Tracing/) package. This includes the `EventSource` and `EventCounter` types, which can be used to emit basic numeric data that can be consumed by external processes like the [dotnet-counters global tool](https://github.com/dotnet/diagnostics/blob/master/documentation/dotnet-counters-instructions.md), or Event Tracing for Windows. To learn more about using `EventCounter` in your own code refer to the [EventCounter Introduction Tutorial](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Tracing/documentation/EventCounterTutorial.md).
+
+For more advanced metrics, and for writing metric data to a wider range of data stores, there is an excellent open source project called [App Metrics](https://www.app-metrics.io). This suite of libraries provides an extensive set of types to instrument your code, and offers packages to write metrics to a variety of targets including time-series databases (e.g. Prometheus and InfluxDB), [Azure AppInsights](https://docs.microsoft.com/azure/azure-monitor/app/app-insights-overview) and more. The [App.Metrics.AspNetCore.Mvc](https://www.nuget.org/packages/App.Metrics.AspNetCore.Mvc/) package even adds a comprehensive set of basic metrics automatically generated via integration with the ASP.NET Core framework, and the web site provides templates for displaying those the [Grafana](https://grafana.com/) visualization platform.
+
+More information and full documentation on AppMetrics is available through their web site at [app-metrics.io](https://app-metrics.io).
```

Comment:
Is there a particular reason you've chosen AppMetrics? I'm not against it, I'm just curious.

As far as I'm aware Application Insights is the only product to offer an SDK (in beta today, will ship with 3.0) that allows you to consume metrics. Unfortunately, the docs haven't yet been updated to reflect that so I don't have something for you to cross-link to.

Reply:
In my experience AppMetrics is the most complete .NET library for doing the full range of metrics and it's back-end agnostic, including supporting AppInsights.